radix-engine-stores/src/hash_tree/mod.rs

```diff
@@ -20,55 +23,289 @@ mod test;
+/// split into "change of ReNode layer's leaf" and "change of substate offset's value in the nested
+/// tree".
+pub struct IdChange<I, V> {
+    /// ID.
+    id: I,
+    /// A value after change, or `None` if this change denotes a delete.
+    changed: Option<V>,
+}
+
+impl<I, V> IdChange<I, V> {
+    pub fn new(id: I, changed: Option<V>) -> Self {
+        Self { id, changed }
+    }
+}
+
+/// A top-level `IdChange`, representing an actual change of a specific substate's hashed value.
+pub type SubstateHashChange = IdChange<SubstateId, Hash>;
+
+/// Inserts a new set of nodes at version `current_version` + 1 into the "nested JMT" persisted
+/// within the given `store`.
+/// In a traditional JMT, this inserts a new leaf node for each given "change", together with an
+/// entire new "parent chain" leading from that leaf to a new root node (common for all of them).
+/// In our instantiation of the JMT, we first update all nested per-`ReNodeModule` trees (i.e. of
+/// each {`RENodeId`, `NodeModuleId`} pair encountered in the `changes`), and then we update the
+/// single upper-layer tree (representing all `ReNodeModule`).
+/// All nodes that became stale precisely due to this (i.e. not any previous) operation will be
+/// reported before the function returns (see `WriteableTreeStore::record_stale_node`).
+/// Returns the hash of the newly-created root (i.e. representing state at version
+/// `current_version` + 1).
 ///
 /// # Panics
-/// Panics if a root node for `current_version` does not exist. The caller
-/// should use `None` to denote an empty, initial state of the tree (i.e.
-/// inserting at version 1).
+/// Panics if a root node for `current_version` does not exist. The caller should use `None` to
+/// denote an empty, initial state of the tree (i.e. inserting at version 1).
 pub fn put_at_next_version<S: TreeStore>(
     store: &mut S,
     current_version: Option<Version>,
-    changes: &[(SubstateId, Option<Hash>)],
+    changes: Vec<SubstateHashChange>,
 ) -> Hash {
-    let value_set: Vec<(Hash, Option<(Hash, SubstateId)>)> = changes
-        .iter()
-        .map(|(id, value_hash)| {
-            (
-                hash(scrypto_encode(id).unwrap()),
-                value_hash.map(|value_hash| (value_hash, id.clone())),
-            )
+    let changes_by_re_node_module = index_by_re_node_module(changes);
+    let mut nested_root_changes = Vec::new();
+    for (re_node_module, substate_changes) in changes_by_re_node_module {
+        let nested_root =
+            put_substate_changes(store, current_version, &re_node_module, substate_changes);
+        nested_root_changes.push(IdChange::new(re_node_module, nested_root));
+    }
+    put_changes(
+        store,
+        current_version,
+        current_version.unwrap_or(0) + 1,
+        nested_root_changes
+            .into_iter()
+            .map(|change| to_re_node_change(change))
+            .collect(),
+    )
+}
+
+// only internals below
```

Comment:
@talekhinezh and @dhedey please let me know if you prefer to have an extensive doc for every struct/function below (it is not easy to tell for me [anymore] which parts are non-obvious, so I documented the public api and left out the internals).